src/Controller/Component/SecurityComponent.php

```diff
@@ -486,6 +490,66 @@ protected function _sortedUnlocked($data)
     }
 
     /**
+     * Get the optional string
+     *
+     * @param array $data Data array
+     * @return array Optional fields and the values they must have if present
+     */
+    protected function _optional(array $data)
+    {
+        if (!empty($data['_Token']['optional'])) {
+            $optional = urldecode($data['_Token']['optional']);
+            $optionalFields = [];
+            if (!empty($optional)) {
+                $optional = explode('|', $optional);
+                foreach ($optional as $field) {
+                    list($name, $hash) = explode('=', $field);
+                    $optionalFields[$name] = $hash;
+                }
+            }
+            return $optionalFields;
+        } else {
+            return [];
+        }
+    }
+
+    /**
+     * Get the sorted optional string
+     *
+     * @param array $data Data array
+     * @return string
+     */
+    protected function _sortedOptional($data)
+    {
+        $optional = $this->_optional($data);
+        ksort($optional, SORT_STRING);
+        return implode('|', array_keys($optional));
+    }
+
+    /**
+     * Validate that any optional fields are either not present or the value matches
+     *
+     * @param \Cake\Controller\Controller $controller Instantiating controller
+     * @param string $token The token string used in hashing
+     * @return bool true if no errors are found
+     */
+    protected function _validateOptionalFields(Controller $controller, $token)
+    {
+        $fields = $controller->request->data;
+        unset($fields['_Token'], $fields['_csrfToken']);
+        $fields = Hash::flatten($fields);
```

Comment:
Any way we could reduce the duplication here and when the required fields are validated?